src/nvim/api/vim.c

```diff
@@ -1468,3 +1474,17 @@ Float nvim__id_float(Float flt)
 {
   return flt;
 }
+
+Array nvim_list_uis(void)
+  FUNC_API_SINCE(1)
+{
+  Array alluis = ARRAY_DICT_INIT;
+  for (unsigned int i = 0; i < ui_count ; i++) {
+    Dictionary dic = ARRAY_DICT_INIT;
+    PUT(dic, "width", INTEGER_OBJ(uis[i]->width));
+    PUT(dic, "height", INTEGER_OBJ(uis[i]->height));
+    PUT(dic, "ui_options", OBJECT_OBJ(api_metadata().items[3].value));
```

Comment:
Metadata is effectively a compile-time constant, instead use a `UIExtension k` loop with `ui_ext_names[k]` and `ui->ui_ext[k]`. Also I think we can just include the options in the top level dict for each UI.

Reply:
okay, if we include the options at the top level dict, should I omit `ui_options` ?

Reply:
Yes.

Reply:
@bfredl I spent quite a lot of time figuring this out but failed,
putting the following inside `nvim_list_uis` does not work because no way 
`uis[i]->ui_ext_names[i]` is valid.
```
for (UIExtension i = 0; i < kUIExtCount; i++) {
  PUT(dic, ui_ext_names[i], copy_object(uis[i]->ui_ext_names[i]));
}
```
I don't know if `ui->ui_ext[k]` can help here, because `ui_ext` is a bool and is a belongs to the `ui`

Either will have to just hardcode the `ui_options` properties one by one else have to use some kind of offset as described here i guess: http://c-faq.com/struct/fieldnames.html

Reply:
Another problem I faced was determining the type of `ui_option`, here `copy_object` will only take in `Objects` what about other structures?

Reply:
`ui->ui_ext[i]` is a bool because the extension options are all bools. So `BOOLEAN_OBJ()` should be used and not `copy_object`. In fact all ui options are currently bool, unless we also count `height` and `width` as options.

Reply:
Yes, I implemented that at first, like this, and thought would not work if `ui_ext_names` had corresponding diff types of types in `ui_ext`, which i now realize does not happen in C :
```c
for (UIExtension j = 0; j < kUIExtCount; j++) {
  PUT(dic, ui_ext_names[j], BOOLEAN_OBJ(uis[i]->ui_ext[j]));
}
```
But I didn't realize corresponding `ui_ext` for `ui_ext_names` had all `bools` as of now. I got confused. This works fine for now. It was so simple!

output is something like this:
```
[{'ext_cmdline': v:false, 'ext_wildmenu': v:false, 'ext_popupmenu': v:false, 'width': 95, 'height': 47, 'ext_tabline': v:false}]
```

I've pushed the changes